File: leetcode_python/Stack/mini-parser.py
"""

385. Mini Parser
Medium

Given a string s represents the serialization of a nested list, implement a parser to deserialize it and return the deserialized NestedInteger.

Each element is either an integer or a list whose elements may also be integers or other lists.

Example 1:

Input: s = "324"
Output: 324
Explanation: You should return a NestedInteger object which contains a single integer 324.

Example 2:

Input: s = "[123,[456,[789]]]"
Output: [123,[456,[789]]]
Explanation: Return a NestedInteger object containing a nested list with 2 elements:
1. An integer containing value 123.
2. A nested list containing two elements:
    i.  An integer containing value 456.
    ii. A nested list with one element:
         a. An integer containing value 789

Constraints:

1 <= s.length <= 5 * 10^4
s consists of digits, square brackets "[]", negative sign '-', and commas ','.
s is the serialization of valid NestedInteger.
All the values in the input are in the range [-10^6, 10^6].

"""

# V0 

# V1
# http://bookshadow.com/weblog/2016/08/15/leetcode-mini-parser/
# time = O(n)
# space = O(n)
class Solution(object):
    def deserialize(self, s):
        """
        :type s: str
        :rtype: NestedInteger
        """
        stack = []
        sigma, negmul = None, 1
        for c in s:
            if c == '-':
                negmul = -1
            elif '0' <= c <= '9':
                sigma = (sigma or 0) * 10 + int(c)
            elif c == '[':
                stack.append(NestedInteger())
            else:
                if sigma is not None:
                    stack[-1].add(NestedInteger(sigma * negmul))
                    sigma, negmul = None, 1
                if c == ']':
                    top = stack.pop()
                    if stack: stack[-1].add(top)
                    else: return top
        return NestedInteger((sigma or 0) * negmul)

# V2
# time = O(n)
# space = O(h)
class NestedInteger(object):
   def __init__(self, value=None):
       """
       If value is not specified, initializes an empty list.
       Otherwise initializes a single integer equal to value.
       """

   def isInteger(self):
       """
       @return True if this NestedInteger holds a single integer, rather than a nested list.
       :rtype bool
       """

   def add(self, elem):
       """
       Set this NestedInteger to hold a nested list and adds a nested integer elem to it.
       :rtype void
       """

   def setInteger(self, value):
       """
       Set this NestedInteger to hold a single integer equal to value.
       :rtype void
       """

   def getInteger(self):
       """
       @return the single integer that this NestedInteger holds, if it holds a single integer
       Return None if this NestedInteger holds a nested list
       :rtype int
       """

   def getList(self):
       """
       @return the nested list that this NestedInteger holds, if it holds a nested list
       Return None if this NestedInteger holds a single integer
       :rtype List[NestedInteger]
       """


# time = O(n)
# space = O(n)
class Solution(object):
    def deserialize(self, s):
        if not s:
            return NestedInteger()

        if s[0] != '[':
            return NestedInteger(int(s))

        stk = []

        i = 0
        for j in xrange(len(s)):
            if s[j] == '[':
                stk += NestedInteger(),
                i = j+1
            elif s[j] in ',]':
                if s[j-1].isdigit():
                    stk[-1].add(NestedInteger(int(s[i:j])))
                if s[j] == ']' and len(stk) > 1:
                    cur = stk[-1]
                    stk.pop()
                    stk[-1].add(cur)
                i = j+1

        return stk[-1]

File: leetcode_java/src/main/java/LeetCodeJava/Stack/MiniParser.java
package LeetCodeJava.Stack;

// https://leetcode.com/problems/mini-parser/

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 *  385. Mini Parser
 *  Medium
 *
 *  Given a string s represents the serialization of a nested list, implement a parser
 *  to deserialize it and return the deserialized NestedInteger.
 *
 *  Each element is either an integer or a list whose elements may also be integers
 *  or other lists.
 *
 *  Example 1:
 *  Input: s = "324"
 *  Output: 324
 *  Explanation: You should return a NestedInteger object which contains a single integer 324.
 *
 *  Example 2:
 *  Input: s = "[123,[456,[789]]]"
 *  Output: [123,[456,[789]]]
 *
 *  Constraints:
 *  1 <= s.length <= 5 * 10^4
 *  s consists of digits, square brackets "[]", negative sign '-', and commas ','.
 *  s is the serialization of valid NestedInteger.
 *  All the values in the input are in the range [-10^6, 10^6].
 */
public class MiniParser {

    /**
     * This is the interface that allows for creating nested lists.
     * On LeetCode it is provided by the judge — reproduced here so the file compiles.
     */
    public static class NestedInteger {

        private Integer value;
        private final List<NestedInteger> list;

        /** Initializes an empty nested list. */
        public NestedInteger() {
            this.value = null;
            this.list = new ArrayList<>();
        }

        /** Initializes a single integer equal to value. */
        public NestedInteger(int value) {
            this.value = value;
            this.list = new ArrayList<>();
        }

        /** @return true if this NestedInteger holds a single integer, rather than a nested list. */
        public boolean isInteger() {
            return this.value != null;
        }

        /** @return the single integer this holds, or null if it holds a nested list. */
        public Integer getInteger() {
            return this.value;
        }

        /** Set this NestedInteger to hold a single integer. */
        public void setInteger(int value) {
            this.value = value;
        }

        /** Set this NestedInteger to hold a nested list and add a nested integer to it. */
        public void add(NestedInteger ni) {
            this.value = null;
            this.list.add(ni);
        }

        /** @return the nested list this holds, or an empty list if it holds a single integer. */
        public List<NestedInteger> getList() {
            return this.list;
        }
    }

    // V0
    // IDEA: STACK — '[' opens a new NestedInteger, ']' closes the current one into its parent,
    //       digits (with optional '-') are parsed as a number and added to the current one
    /**
     * time = O(n)
     * space = O(d)   // d = max nesting depth
     */
    public NestedInteger deserialize(String s) {

        if (s == null || s.isEmpty()) {
            return new NestedInteger();
        }

        // a bare number, no brackets at all
        if (s.charAt(0) != '[') {
            return new NestedInteger(Integer.parseInt(s));
        }

        Deque<NestedInteger> stack = new ArrayDeque<>();
        NestedInteger cur = null;
        int i = 0;
        while (i < s.length()) {
            char c = s.charAt(i);
            if (c == '[') {
                if (cur != null) {
                    stack.push(cur);
                }
                cur = new NestedInteger();
                i++;
            } else if (c == ']') {
                if (!stack.isEmpty()) {
                    NestedInteger parent = stack.pop();
                    parent.add(cur);
                    cur = parent;
                }
                i++;
            } else if (c == ',') {
                i++;
            } else {
                // read a (possibly negative) integer
                int j = i;
                if (s.charAt(j) == '-') {
                    j++;
                }
                while (j < s.length() && Character.isDigit(s.charAt(j))) {
                    j++;
                }
                cur.add(new NestedInteger(Integer.parseInt(s.substring(i, j))));
                i = j;
            }
        }
        return cur;
    }
}
